Bug 3390: Proxy auth data visible to scripts

yadij opened the pull request:
Original changes to redact credentials from error page %R code
expansion output was incomplete. It missed the parse failure
case where ErrorState::request_hdrs raw buffer contained
sensitive information.

Also missed was the %W case where full request message headers
were generated in a mailto link. This case is especially
problematic as it may be delivered over insecure SMTP even if
the error was secured with HTTPS.

After this change:
* the HttpRequest message packing code for error pages is
  de-duplicated and elides authentication headers for both %R
  and %W code outputs.
* The %W code displays the request message protocol label
   correctly.
* The %R code output includes the CRLF request message
   terminator.
* The email_err_data directive causing advanced details to be
   added to %W mailto links is disabled by default.

yadij commented on src/HttpRequest.cc:
```diff
@@ -471,9 +471,10 @@ HttpRequest::packFirstLineInto(Packable * p, bool full_uri) const
+    p->appendf(SQUIDSBUFPH " " SQUIDSBUFPH " %s/%d.%d\r\n",
                SQUIDSBUFPRINT(method.image()),
                SQUIDSBUFPRINT(tmp),
+               AnyP::ProtocolType_str[http_ver.protocol],
```
FTR: previous bug in ICY and ICAP error messages %W output. Now fixed to ensure identical %W/%R logic.

rousskov commented on src/errorpage.cc:
```diff
@@ -1149,18 +1148,10 @@ ErrorState::compileLegacyCode(Build &build)
-            p = request_hdrs;
-        } else {
+        else if (request)
+            request->pack(&mb, true /* hide authorization data */);
```
This will append CRLF that was not appended before this changes. I hope that is OK, but perhaps this PR description should mention that `%R` output has changed a little.

yadij replied:
Not sure documenting this is truly helpful.

Error pages contain ```<pre>%R</pre>```. Which goes from this:
```
   <pre>GET url HTTP/1.1
header
header</pre>
```
to this:
```
   <pre>GET url HTTP/1.1
header
header
</pre>
```

yadij replied:
added a line to description anyway.

rousskov reviewed (changes requested):
Thank you for working on this bug fix!

rousskov commented:
> RFC: should we add `Cookie:` and `Cookie2:` header redaction now?

I would support removal of any header others might consider "sensitive", but please do that in a dedicated PR.
 
> IIRC, there have been a small number of requests to do so over the years.

FWIW, I think the whole idea of echoing requests should be retired. It was a great idea in 1997! And, yes, it _is_ still useful in certain cases, but it is not useful enough overall to keep risking so much (and to keep spending our scarce development resources on). Instead, we should report the equivalent of a failed master transaction ID (and spend our resources on improving transaction ID tracking).

yadij commented on src/errorpage.cc:
```diff
@@ -1149,18 +1148,10 @@ ErrorState::compileLegacyCode(Build &build)
-            p = request_hdrs;
-        } else {
+        else if (request)
+            request->pack(&mb, true /* hide authorization data */);
```
The difference between these cases is that we can reliably redact when the parse succeeded, we cannot reliably do so when faced with opaque buffer content. For this patch the minimum reasonable redaction is performed.

rousskov replied:
I disagree that Squid can "reliably redact" HTTP request headers (without admin help). Yes, Squid can automatically delete a few sensitive headers we know about, but there may be other sensitive information in those headers that Squid will echo back (unless we stop echoing request headers).

I disagree that the chosen redaction is enough to address the problem. In other words, the chosen code changes are less than the required minimum.

kinkie replied:
> the chosen code changes are less than the required minimum.

Maybe. This can be addressed in a followup PR as needed

rousskov replied:
> > the chosen code changes are less than the required minimum.

> Maybe. This can be addressed in a followup PR as needed

Merging a PR that does less than the required minimum feels illogical to me when addressing this particular concern in this particular PR is trivial.

yadij commented:
> I reworked "ASCII prefix of raw buffer content is no longer available for output in custom error pages" bullet 

That level of details is not desirable. Removed it entirely.

> AFAICT, we could even dump raw HTTP body (i.e. past-header) bytes if we successfully parsed HTTP request but failed to parse its target!

... ah "target" is is part of the request-line. Did you mean "content"?

We cannot know whether that is shorter than the buffer and another request header with sensitive data is later in the raw/opaque input.

Also, these errors are not about the body/content/payload. Displaying any of it is not wanted to begin with.

> 
> For unparseable HTTP request _headers_, I hope the following ConnStateData::parseHttpRequest() code empties that raw input buffer, but I am not 100% sure there are no other ways to reach Http::One::Server::buildHttpRequest() with some unparseable request header bytes in inBuf. I have not checked FTP path at all.
> 

I'm not sure why you are focusing on this action. When the `inBuf` is no longer useful the entire connection and all state related to it is about to be closed at the TCP level as soon as the error write completes. No need to do extra work AFAIK.

rousskov reviewed (commented):
> @rousskov the entire purpose of these page is for human (client/user) use. Your request to elide the debug details is a paradigm shift to producing machine-focused errors instead of human-focused. Which I am calling out of scope.

I have not requested a "shift to producing machine-focused errors instead of human-focused". I have not requested to add any machine-focused information. I requested to _remove_ echoing of parsed request headers from error pages. Such a _removal_ does not shift error pages focus and ought to be in this PR scope (even if one believes that HTTP header dump can be called "human-focused" when presented to a typical human user browsing the web).

> > AFAICT, we could even dump raw HTTP body (i.e. past-header) bytes if we successfully parsed HTTP request but failed to parse its target!

> ... ah "target" is is part of the request-line. Did you mean "content"?

No, I did not. I meant [request target](https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2) (a.k.a. request URI), which is a part of the request-line that we (re)parse/interpret later/separately from the request prefix. See a FromUrlXXX() call inside Http::One::Server::buildHttpRequest().
 
> > For unparseable HTTP request _headers_, I hope the following ConnStateData::parseHttpRequest() code empties that raw input buffer, but I am not 100% sure there are no other ways to reach Http::One::Server::buildHttpRequest() with some unparseable request header bytes in inBuf. I have not checked FTP path at all.

> I'm not sure why you are focusing on this action.

My [comment](https://github.com/squid-cache/squid/pull/2249#pullrequestreview-3234790678) was about an "ASCII prefix" assertion in an earlier PR description. The contents of the previously echoed raw prefix when this bug is triggered using bad request targets depends, in part, on inBuf content. That content may contain request body and/or another request header in certain cases, for example.

> When the `inBuf` is no longer useful the entire connection and all state related to it is about to be closed at the TCP level as soon as the error write completes. No need to do extra work AFAIK.

I believe my PR description edits have addressed the problem my comment was discussing. I have not requested any extra work related to inBuf.

kinkie commented on src/HttpRequest.cc:
```diff
@@ -471,9 +471,10 @@ HttpRequest::packFirstLineInto(Packable * p, bool full_uri) const
 
     // form HTTP request-line
-    p->appendf(SQUIDSBUFPH " " SQUIDSBUFPH " HTTP/%d.%d\r\n",
+    p->appendf(SQUIDSBUFPH " " SQUIDSBUFPH " %s/%d.%d\r\n",
```
Is there any reason to not use stream operators instead here?

yadij commented on src/client_side_reply.cc:
```diff
@@ -1005,11 +1002,15 @@ clientReplyContext::traceReply()
     const HttpReplyPointer rep(new HttpReply);
-    rep->setHeaders(Http::scOkay, nullptr, "text/plain", http->request->prefixLen(), 0, squid_curtime);
+    rep->setHeaders(Http::scOkay, nullptr, "message/http", content.contentSize(), 0, squid_curtime);
+    rep->body.set(SBuf(content.buf, content.size));
```
This is where the non-filters vs filtered body is determined. It is a fixed known-length and the `HttpReply::body` member is documented as existign for this type of use-case. So setting up `clientStreams` for handling the body seems too much complexity.

After your last review I was retaining the `swapOut()` under the impression that it was the trigger starting the delivery process. My testing did not show any excess data being delivered so I assumed (incorrectly) that it was the right thing to do. A test without it has show no difference, so dropped in [dba94b2](https://github.com/squid-cache/squid/pull/2249/commits/dba94b280a7dde5fbe0b93f6fff2943b95edee63)

yadij commented on src/client_side_reply.cc:
```diff
@@ -1005,11 +1002,14 @@ clientReplyContext::traceReply()
+    rep->setHeaders(Http::scOkay, nullptr, "message/http", content.contentSize(), 0, squid_curtime);
+    rep->body.set(SBuf(content.buf, content.size));
     http->storeEntry()->replaceHttpReply(rep);
-    http->request->swapOut(http->storeEntry());
```
will do as a followup.

rousskov reviewed (commented):
I removed a stale item from PR description:

* The %W code displays the request message protocol label correctly.
